### ISSUE TITLE
keep_last_value not working correctly for sparse data

### ISSUE THREAD

**Issue by matejsp** [VictoriaMetrics/VictoriaMetrics#1411]
There is a problem when using keep_last_value and getting different results (correct, empty or partial result) when changing step when sparse data exists in the database. Expected result would be if I change the step I should get the same data (with different granulation of course).

Example:

I have written 3 history points (on 12.4.2021) to VictoriaMetrics (1.61.1):

http://localhost:8/api/v1/export?match[]=asset_balance

```
{"metric":{"__name__":"asset_balance","currency":"EUR"},"values":[5000,0],"timestamps":[1618234254000,1618235399000]}
{"metric":{"__name__":"asset_balance","currency":"TEST"},"values":[654.835956],"timestamps":[1618235399000]}
```

I am selecting in Grafana:
- time range -> 2021-04-10 00:00:00 -> 2021-04-15 00:00:00
- query -> keep_last_value(asset_balance):

step 5m 2 series returned (only correct result)

![image](https://user-images.githubusercontent.com/3091122/123647758-a5fc5100-d828-11eb-999d-ad13507492b1.png)

step 15m: 1 series returned one missing.

![image](https://user-images.githubusercontent.com/3091122/123649956-73535800-d82a-11eb-9800-690ce85c744f.png)

step 1h: 1 series returned one missing.

![image](https://user-images.githubusercontent.com/3091122/123647659-91b85400-d828-11eb-8032-32448b14a1ff.png)

step 1d: 0 series returned. two missing

![image](https://user-images.githubusercontent.com/3091122/123647484-66ce0000-d828-11eb-9d1e-0c01deab9927.png)

Any ideas how to solve this?

**Comment by valyala** [comment]
`keep_last_value(q)` works on samples returned from the inner query `q`. If `q` doesn't return any samples, then `keep_last_value()` won't return any samples too. E.g. try running the query `asset_balance` with different `step` values and see it returns only a few points for steps 15m and 1h, while it returns zero points for `step=1d`. This is due to Prometheus data model:
- it expects that each time series contain samples at regular intervals called `scrape_interval
- it returns up to `1 + (end - start) / step` points with `step` distance between them for each series from [/api/v1/query_range](https://prometheus.io/docs/prometheus/latest/querying/api/#range-queries) handler. It returns the last real sample value for every point on the graph. It stops returning the last sample value if it is located further than `2x scrape_interval` from the current point.

See more details at https://prometheus.io/docs/prometheus/latest/querying/basics/#staleness .

Try using `last_over_time(m[d])` function instead of `keep_last_value(m)`. The `last_over_time(m[d])` returns the last real sample value per each series over the duration `d`. If `[d]` is missing, e.g. if `last_over_time(asset_balance)` is used, then VictoriaMetrics automatically sets `d` to the specified `step` value.

**Comment by matejsp** [comment]
You are correct. keep_last_value was not the problem. last_over_time is indeed fixing my problem.

Example:
```
keep_last_value(last_over_time(asset_balance)) for step 1d is now returning me the expected result
```

In prometheus scrape_interval is configurable per scrape. I see that victoria has:
 -search.minStalenessInterval
 -search.maxStalenessInterval
but it is not configurable per query (series)? If I understand correctly I should set maxStalnessInterval to --retentionPeriod? What is the meaning of min staleness? Is victoria skipping points if more points than min interval?

> If the gaps are related to irregular intervals between samples, then try adjusting -search.minStalenessInterval command-line flag to value close to the maximum interval between samples.
So I should set both to retentionPeriod and that would be it.

> By default it is automatically calculated from the median interval between samples
Is this calculated for each series or globally or even across tenants?

Is staleness actually an issue for victoria data model or just prometheus feature parity? Since here each point has timestamp present. Is it affecting performance? I wouldn't replicate "information loss" feature that is advertised as Gothas in prometheus.

Can you please update documentation a bit where makes sense. I mean those gothas in prometheus are documented in query/basics section but in victoria they are a lot more hidden :)

We migrated from influxdb so this prometheus model is a bit wierd to me :)

**Comment by valyala** [comment]
> keep_last_value(last_over_time(asset_balance))

The query can be simplified to `last_over_time(asset_balance)`. There is no need in wrapping it into `keep_last_value()`.

> I see that victoria has:
-search.minStalenessInterval
-search.maxStalenessInterval
but it is not configurable per query (series)?

The `-search.maxStalenessInterval` can be configured per query via `max_lookback` query param passed to [/api/v1/query](https://prometheus.io/docs/prometheus/latest/querying/api/#instant-queries) and to [/api/v1/query_range](https://prometheus.io/docs/prometheus/latest/querying/api/#range-queries). For example, `/api/v1/query?query=last_over_time(asset_balance)&max_lookback=5m` would emulate Prometheus behaviour on staleness detection, e.g. the maximum staleness interval will be limited by 5 minutes.

The `-search.minStalenessInterval` can be emulated with `last_over_time()` function on a per-query basis. For example, `last_over_time(asset_balance[1d])` would return the same results as if VictoriaMetrics is configured with `-search.minStalenessInterval=1d` command-line flag, e.g. it continues returning the last raw sample value for one day after the raw sample timestamp.

> If I understand correctly I should set maxStalnessInterval to --retentionPeriod?

No. This doesn't make sense in general case.

> What is the meaning of min staleness?

This means that VictoriaMetrics continues returning non-empty points for up to `-search.minStalenessInterval` after the last raw sample stored in the database. E.g. this is equivalent to `last_over_time(metric[d])` if `d` is passed to `-search.minStalenessInterval=d`.


> Is staleness actually an issue for victoria data model or just prometheus feature parity?

The staleness detection works differently in Prometheus and VictoriaMetrics. Prometheus stops returning points from `/api/v1/query` and `/api/v1/query_range` handlers if time distance between the returned point and the previous raw sample exceeds 5 minutes. VictoriaMetrics stops returning points if the time distance between the returned point and the previous raw samples exceeds `2 x scrape_interval`, where `scrape_interval` is automatically detected as the median interval between adjacent samples in a time series.

> Can you please update documentation a bit where makes sense. I mean those gothas in prometheus are documented in query/basics section but in victoria they are a lot more hidden :)

I think it would be great providing a separate docs like `Migrating from InfluxDB to VictoriaMetrics` with the description of differences and gotchas in data models between InfluxDB and VictoriaMetrics.

**Comment by valyala** [comment]
FYI, we created a guide on [how to migrate from InfluxDB to VictoriaMetrics](https://docs.victoriametrics.com/guides/migrate-from-influx/).

Closing the issue as answered.

**Comment by matejsp** [comment]
@valyala I have read the guide and it is very nice. One thing thar I noticed that is not in the guide but can be important especially during migration.

Basically metrics in victoria are off by one when you are grouping by time.

10:01:00
bucket 1
10:02:00
bucket 2
10:03:00

So when you use sum_over_time[1m] vs group by 1m in influxdb you are getting:
10:02:00 -> bucket 2 -> for infludb
10:03:00 -> bucket 2 -> for victoriadb

The points are stored ok, just what time is used for grouped timestamp (first point for range vs last point for range).
So to migrate we had to do sum_over_time(1m) - 1m to get the exact same result.
